R/check_reliability.R

```diff
@@ -0,0 +1,108 @@
+#' Variability-Over-Uncertainty Ratio (D-vour) for Random Effects Reliability
+#'
+#' @description TODO: Add description.
+#'
+#' @param x A model object.
+#' @param ... Currently not used.
+#'
+#
+#'
+#' @details TODO: Add details.
+#'
+#' @references TODO.
+#'
+#' @family functions to check model assumptions and and assess model quality
+#'
+#' @examplesIf require("lme4")
+#' # Add groups to the data
+#' data <- iris
+#' data$Place <- as.factor(rep(c("P1", "P2", "P3", "P4", "P5", "P6"), each = 25))
+#'
+#' # lme4
+#' m <- lme4::lmer(Sepal.Width ~ Petal.Width + (Petal.Width | Place), data = data)
+#' check_reliability(m)
+#'
+#' @export
+check_reliability <- function(x, ...) {
+  UseMethod("check_reliability")
+}
+
+
+#' @export
+check_reliability.default <- function(x, ...) {
+  insight::check_if_installed("modelbased", minimum_version = "0.10.0")
+  check_reliability(modelbased::estimate_grouplevel(x, ...), ...)
+}
+
+
+#' @export
+check_reliability.estimate_grouplevel <- function(x, ...) {
+  coefname <- attributes(x)$coef_name
+  dispname <- grep("SE|SD|MAD", colnames(x), value = TRUE)
+
+  # Sanity checks
+  if (insight::n_unique(x$Level) <= 3) {
+    insight::format_alert(paste0(
+      "The number of random effects group levels (N=",
+      insight::n_unique(x$Level),
+      ") might be too low to reliably estimate the variability."
+    ))
+  }
+
+  if (length(dispname) == 0) {
+    insight::format_error(paste0(
+      "This function requires an index of variability of each random ",
+      "effect (e.g., SE) but none was found. Try running `check_reliability()` on the ",
+      "output of `modelbased::estimate_grouplevel(model)`, and make sure the latter ",
+      "returns a table with an index of dispersion."
+    ))
+  }
+
+  if (length(dispname) > 1) {
+    insight::format_alert(paste0(
+      "Multiple indices of variability were found (",
+      toString(dispname),
+      "). Using the first one."
+    ))
+    dispname <- dispname[1]
+  }
+
+  # Compute reliability
+  if (!"Component" %in% names(x)) x$Component <- "TEMP"
+
+  reliability <- data.frame()
+  for (comp in unique(x$Component)) {
+    for (grp in unique(x$Group)) {
+      for (param in unique(x$Parameter)) {
+        d <- x[x$Component == comp & x$Group == grp & x$Parameter == param, ]
+
+        # Raw Signal-to-Noise Ratio
+        rez <- data.frame(
+          Component = comp,
+          Group = grp,
+          Parameter = param,
+          Variability = stats::sd(d[[coefname]]),
+          Uncertainty = mean(d[[dispname]])
+        )
+        rez$Reliability <- rez$Variability / rez$Uncertainty
+
+        # Alternative: average of level-specific reliability
+        rez$Reliability2 <- mean(d[[coefname]]^2 / (d[[coefname]]^2 + d[[dispname]]^2))
+
+        # Alternative 2: like hlmer?
+        rez$Reliability3 <- rez$Variability^2 / (rez$Variability^2 + rez$Uncertainty^2)
```

Comment:
Grok said:

The suggestion is to compute reliability as coefs[i]^2 / (coefs[i]^2 + SEs[i]^2) for each level (cluster) of the random effect, then average these values for each parameter (e.g., (Intercept) and Petal.Width). This:
- Uses squared coefficients (coefs[i]^2) and squared standard errors (SEs[i]^2) to work in variance-like units, consistent with hlmer.
- Calculates a reliability ratio per level, then averages across levels, mirroring hlmer’s cluster-level approach.

But indeed, I don't know whether it's better to reliability per level **then** average or average first 🤷

Reply:
> Grok 

You're on the loser's road... 😬 

Given that the reliability index is the ratio of estimate variability to error variability, SD^2 seems to make more sense (see reliability3):

m <- insight::download_model("brms_mixed_10")
x <- modelbased::estimate_grouplevel(m)

performance::check_reliability(m)
#> The number of random effects group levels (N=3) might be too low to
#>   reliably estimate the variability.
#>   Group   Parameter Variability Uncertainty Reliability Reliability2
#> 1 Group   Intercept  0.92950881   0.8448505    1.100205    0.3569732
#> 2 Group Petal.Width  0.04800641   0.1873084    0.256296    0.0467122
#>   Reliability3
#> 1   0.54760370
#> 2   0.06163875